base/subarray.jl

```diff
@@ -100,90 +128,61 @@ end
-    (@_propagate_inbounds_meta; (merge_indexes(V, idxs, subidxs[1]),))
-# As an optimization, we don't need to merge indices if all trailing indices are dropped scalars
-reindex(V, idxs::Tuple{Colon, Vararg{DroppedScalar}}, subidxs::Tuple{Any}) =
-    (@_propagate_inbounds_meta; (subidxs[1], tail(idxs)...))
 
 # Re-index into parent vectors with one subindex
-reindex(V, idxs::Tuple{AbstractVector}, subidxs::Tuple{Any}) =
-    (@_propagate_inbounds_meta; (idxs[1][subidxs[1]],))
 reindex(V, idxs::Tuple{AbstractVector, Vararg{Any}}, subidxs::Tuple{Any, Vararg{Any}}) =
     (@_propagate_inbounds_meta; (idxs[1][subidxs[1]], reindex(V, tail(idxs), tail(subidxs))...))
-reindex(V, idxs::Tuple{AbstractVector, Vararg{Any}}, subidxs::Tuple{Any}) =
-    (@_propagate_inbounds_meta; (merge_indexes(V, idxs, subidxs[1]),))
-reindex(V, idxs::Tuple{AbstractVector, Vararg{DroppedScalar}}, subidxs::Tuple{Any}) =
-    (@_propagate_inbounds_meta; (idxs[1][subidxs[1]], tail(idxs)...))
 
 # Parent matrices are re-indexed with two sub-indices
-reindex(V, idxs::Tuple{AbstractMatrix}, subidxs::Tuple{Any}) =
-    (@_propagate_inbounds_meta; (idxs[1][subidxs[1]],))
-reindex(V, idxs::Tuple{AbstractMatrix}, subidxs::Tuple{Any, Any}) =
-    (@_propagate_inbounds_meta; (idxs[1][subidxs[1], subidxs[2]],))
 reindex(V, idxs::Tuple{AbstractMatrix, Vararg{Any}}, subidxs::Tuple{Any, Any, Vararg{Any}}) =
     (@_propagate_inbounds_meta; (idxs[1][subidxs[1], subidxs[2]], reindex(V, tail(idxs), tail(tail(subidxs)))...))
-reindex(V, idxs::Tuple{AbstractMatrix, Vararg{Any}}, subidxs::Tuple{Any}) =
-    (@_propagate_inbounds_meta; (merge_indexes(V, idxs, subidxs[1]),))
-reindex(V, idxs::Tuple{AbstractMatrix, Vararg{Any}}, subidxs::Tuple{Any, Any}) =
-    (@_propagate_inbounds_meta; (merge_indexes(V, idxs, subidxs),))
-reindex(V, idxs::Tuple{AbstractMatrix, Vararg{DroppedScalar}}, subidxs::Tuple{Any}) =
-    (@_propagate_inbounds_meta; (idxs[1][subidxs[1]], tail(idxs)...))
-reindex(V, idxs::Tuple{AbstractMatrix, Vararg{DroppedScalar}}, subidxs::Tuple{Any, Any}) =
-    (@_propagate_inbounds_meta; (idxs[1][subidxs[1], subidxs[2]], tail(idxs)...))
 
 # In general, we index N-dimensional parent arrays with N indices
 @generated function reindex{T,N}(V, idxs::Tuple{AbstractArray{T,N}, Vararg{Any}}, subidxs::Tuple{Vararg{Any}})
     if length(subidxs.parameters) >= N
         subs = [:(subidxs[$d]) for d in 1:N]
         tail = [:(subidxs[$d]) for d in N+1:length(subidxs.parameters)]
         :(@_propagate_inbounds_meta; (idxs[1][$(subs...)], reindex(V, tail(idxs), ($(tail...),))...))
-    elseif length(idxs.parameters) == 1
-        :(@_propagate_inbounds_meta; (idxs[1][subidxs...],))
-    elseif all(T->T<:DroppedScalar, idxs.parameters[2:end])
-        :(@_propagate_inbounds_meta; (idxs[1][subidxs...], tail(idxs)...))
     else
-        :(@_propagate_inbounds_meta; (merge_indexes(V, idxs, subidxs),))
+        :(error("mismatched index dimensionalities"))
```

Comment:
Yeah, I should probably enforce this in the constructor.  The whole point here is to ensure that the number of indices stored in the SubArray matches the number of dimensions in the parent array.  Even if you do manage to create a corrupted SubArray, though, you'd almost certainly hit a missing method in the other `reshape` signatures long before you ever get here.

We've typically stayed away from using the word rank to mean `ndims` because of [`rank`](http://docs.julialang.org/en/latest/stdlib/linalg/#Base.rank).